lib/fabrik/DirectorInstance.js

```diff
@@ -5,21 +5,38 @@ const Promise = require('bluebird');
 const _ = require('lodash');
 const yaml = require('js-yaml');
 const BaseInstance = require('./BaseInstance');
+const CfPlatformManager = require('./CfPlatformManager');
+const K8sPlatformManager = require('./K8sPlatformManager');
 const config = require('../config');
 const logger = require('../logger');
 const errors = require('../errors');
 const jwt = require('../jwt');
 const utils = require('../utils');
 const NotFound = errors.NotFound;
+const BadRequest = errors.BadRequest;
 const ServiceInstanceNotFound = errors.ServiceInstanceNotFound;
 const SecurityGroupNotCreated = errors.SecurityGroupNotCreated;
 const ScheduleManager = require('../jobs');
 const CONST = require('../constants');
 
 class DirectorInstance extends BaseInstance {
-  constructor(guid, manager) {
+  constructor(guid, manager, platformManager) {
     super(guid, manager);
     this.networkSegmentIndex = undefined;
+    this.platformManager = !platformManager ? undefined : platformManager;
+  }
+
+  assignPlatformManager() {
```

Comment:
code smell. Why is this `assignPlatformManager` being called from outside? This class's constructor is also passed with platformmanager as a param. Code flow of assigning platform manager seems convoluted.

Rather i would expect the context to be always present & set in req object. If context is absent, you fetch it either from Deployment or Docker environment and ensure its present. Once its existence is guaranteed in flow, just pass it as a param to Instance's constructor method. The respective instances can themselves call `BasePlatformManager.getInstance` and get an instance. 

This way of assigning platformmanager either via constructor or via this explicit method seems redundant. In all calls of POST/UPDATE context will always be present so even in such scenarios unnecessarily calls will be made to BOSH which could've been avoided.

Reply:
'In all calls of POST/UPDATE context will always be present so even in such scenarios unnecessarily calls will be made to BOSH which could've been avoided.' ==> whenever context is present in request body, not extra call is made to BOSH or Docker. Platform manager is built directly based on the value of context.platform.

Reply:
'If context is absent, you fetch it either from Deployment or Docker environment and ensure its present. Once its existence is guaranteed in flow, just pass it as a param to Instance's constructor method.' ==> To determine the docker environment we need to instantiate DocketInstance itself first. So, that's why defined the method to fetch context details.